src/delegate.ts

```diff
@@ -0,0 +1,38 @@
+import { Handle } from 'dojo-core/interfaces';
+import on, { ExtensionEvent } from 'dojo-core/on';
+
+export default function delegate(target: HTMLElement, selector: string, type: string, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: ExtensionEvent, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: (string | ExtensionEvent)[], listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: any, listener: (event: UIEvent) => void): Handle {
+	function matches(target: EventTarget) {
+		let node = <any> target;
```

Comment:
Unfortunately, for some reason, TypeScript's typings include `webkitMatchesSelector` and `msMatchesSelector` in `Element`, but not `matches` (the actual standard).  So at the very least, `<any>` needs to be used for that.  However, in this location it's actually hiding a number of other issues, including the fact that this function is returning either a boolean or a Node, which seems pretty awkward.

I came up with the following instead:

```ts
	function matches(target: HTMLElement) {
		let node = target;
		let matchMethod = (function() {
			// TS typings include the vendor-prefixed methods used here, but not the standard
			if (typeof (<any> node).matches === 'function') {
				return 'matches';
			}
			if (typeof node.webkitMatchesSelector === 'function') {
				return 'webkitMatchesSelector';
			}
			if (typeof node.msMatchesSelector === 'function') {
				return 'msMatchesSelector';
			}
		})();

		// TS7017
		while (!(<any> node)[matchMethod](selector)) {
			if (!node.parentNode || !(<any> node.parentNode)[matchMethod]) {
				return null;
			}
			node = <HTMLElement> node.parentNode;
		}

		return node;
	}

	return on(<EventTarget> target, type, function (event: Event) {
		let matchedEventTarget = matches(<HTMLElement> event.target);
		if (matchedEventTarget) {
			return listener.call(matchedEventTarget, event);
		}
	});
```